### ISSUE TITLE
DOC How is randomization handled

### ISSUE THREAD

**Issue by adrinjalali** [fairlearn/fairlearn#243]
The docs say:

> Randomization. In contrast with scikit-learn, estimators in fairlearn can produce randomized predictors. Randomization of predictions is required to satisfy many definitions of fairness. Because of randomization, it is possible to get different outputs from the predictor's predict method on identical data. For each of our methods, we provide explicit access to the probability distribution used for randomization.

`sklearn` does have randomization in many estimators (`RandomForests` as an example :P). But randomness is always controlled by a `random_seed` parameter. Reproducibility requires setting this parameter, to be able to go back and reproduce the results.

It is understandable if in the context of fairness the RNG shouldn't be fixed, but shouldn't the user be able to feed in a seed or a seed and have reproducible results?

Also, the user can set the RNG, and still get probabilistic output given the same input. I could have:

``` python
clf = MyClassifier(random_seed=42)
clf.fit(X, y)
clf.predict(x0) -> returns 0
clf.predict(x0) -> returns 1
```

but if the user runs the same script again, they'll get the same output as before.

**Comment by romanlutz** [comment]
Your observation is correct. This is somewhat related to https://github.com/fairlearn/fairlearn/issues/65, since this needs to be crystal clear to users, but we've discussed providing several options a few times. There should be an option to set the seed. That may break fairness guarantees, though, so we need to make sure to document this properly.

**Comment by adrinjalali** [comment]
I'm still not sure how setting a seed would break the fairness guarantee here, since the output would still depend on how many `predict`s the user would have called before it gets to this particular input. Also, you can let the users set the seed, but be clear in the docs that in production they shouldn't (if that's the case).

But even then, I'm not sure how the user is supposed to handle the lack of reproducibility of the results.

**Comment by riedgar-ms** [comment]
The lack of reproducibility is a concern of ours. It's a major motivation for us to get customer feedback on the exponentiated gradient algorithm in the Real World(tm). As @romanlutz says, we have been talking about the best way to expose the PRNG to the user to get reproducible results under 'lab' conditions at least.

**Comment by romanlutz** [comment]
What I meant by breaking fairness guarantees is that if you provide the same seed all the time you've effectively derandomized the mitigated predictor. Some of these methods assume that randomization is allowed and don't achieve their objective if that's violated.

What you are saying is that you set the seed once in the beginning only, which is different from what I meant, and shouldn't break the guarantees.

I'll have to think about a good way of handling this and will propose a solution sometime next week (hopefully).

**Comment by adrinjalali** [comment]
> What you are saying is that you set the seed once in the beginning only, which is different from what I meant, and shouldn't break the guarantees.

The scope of the rng at least in sklearn is the model's lifetime, and not just a method call. If you want to pass sklearn's common tests, you kinda have to have it as a constructor parameter.

**Comment by riedgar-ms** [comment]
I expect it would be. The RNG for predict would live for as long as the object. Each new incoming predict request (and if it's an array, each column of the array) needs a new random number, and you don't want to be reseeding it.

**Comment by adrinjalali** [comment]
I also found this one in our docs https://scikit-learn.org/stable/developers/develop.html#random-numbers

**Comment by MiroDudik** [comment]
My original idea was to augment predict, so that we could call, e.g.,:
expgrad.predict(X, random_state=42)

This call format could co-exist with the sklearn convention. The idea is that if the predict method receives the random_state argument, it uses the RandomState object based on that argument rather than the one from expgrad.random_state_. In that case, expgrad.random_state_ does not advance.

**Comment by romanlutz** [comment]
@MiroDudik I'm not particularly opposed to that, but I can't think of a scenario where it's useful. Can you elaborate?

**Comment by MiroDudik** [comment]
One simple scenario is the following:
- you are running a jupyter notebook
- you fit a mitigated estimator in one jupyter cell, it takes a while, but completes without any issues
- in the next jupyter cell, you call predict on the fitted estimator, but there is a bug in printing some kind of a statistic on the predictions
- you fix the bug and re-run the cell, but if the predictions are different, you cannot be quite sure whether you fixed the bug (which was somewhere in calculating the summary of predictions)

More generally--whenever there is any kind of a pipeline or additional processing / analysis done after you fit an estimator, I think that you would may want to replicate the behavior on the level of "predict". Debugging is one use-case. Comparison of two strategies for postprocessing predictions might be another use-case. (You want to compare postprocessing of the same predictions.)

**Comment by adrinjalali** [comment]
That's interesting. I'd say that's a case where a [advanced] user could capture the state of the rng and reset it before calling the predict again.

IMHO passing the rng or a seed to predict is pretty dangerous and may end up with people using it in a way they really shouldn't.

In the scenario you mention, you can still go back and re-run the whole pipeline and get the same result, if you could set the rng at the estimator level, and if you want to re-run a predict, you can set the seed from outside the estimator to force the re-run of the same path.

In contrast to your proposed API, if users set the rng at the estimator level, they will get two different outputs if they call `predict` on the same input, and still they have the reproducibility guarantees.

Also, note that not exposing the rng through the `__init__` doesn't mean it's really random. It means if the user needs to reproduce their results, they should set the global `numpy` seed or whichever rng you use internally in the estimator.

**Comment by MiroDudik** [comment]
Good point re. the ability to just explicitly reset the random seed before the call to predict. And I can also see how people might use the predict-level seed the wrong way.

How about this. Let's go with the `sklearn` convention, i.e., the `random_state` passed in `__init__`. If we ever need the reproducibility on the `predict` level, we can just explicitly reset the random seed on the rng being used. And if this comes up as a repeated issue, we can reconsider introducing `random_seed` override on the `predict` level.

**Comment by riedgar-ms** [comment]
So long as we're very sure that in `__init__` we take a reference to the PRNG, and not a copy.

**Comment by adrinjalali** [comment]
in `sklearn` we never touch any of the given parameters in `__init__` anyway.

Regarding taking the reference, that's how it's done in [`check_random_state`](https://github.com/scikit-learn/scikit-learn/blob/master/sklearn/utils/validation.py#L773) if the users passes a numpy `RandomState`.

The discussion around this may also interest you: https://github.com/scikit-learn/enhancement_proposals/pull/24

**Comment by hildeweerts** [comment]
Is this still relevant? If so, it would be great if someone can either list the steps we need to take to fix this issue (or maybe open a fresh issue). @MiroDudik @adrinjalali @romanlutz @riedgar-ms

**Comment by adrinjalali** [comment]
Just a note that you could also do `@fairlearn/fairlearn-maintainers` instead of mentioning people one by one if you want to ping all maintainers ;) Typing at-sign-fairlearn would give you the autocomplete for the team's name.

**Comment by hildeweerts** [comment]
Oh, I didn't know that. Thanks!

(I actually tried tagging people who engaged in the discussion for each issue, which accidentally mostly coincided with most of the current maintainers :P ).

**Comment by TamaraAtanasoska** [comment]
@adrinjalali I'd love some expert thoughts on this too, especially regarding sklearn changes in the meantime so we know how it makes sense to proceed.

**Comment by adrinjalali** [comment]
The discussion on the sklearn side hasn't moved forward, since the person championing it has left.

But in the meantime, with metadata routing, it's now easy to pass things to methods such as `predict`. So if we want to let users pass an rng to `predict`, we can totally do so.
